permissions/src/main/java/com/google/accompanist/permissions/PermissionsRequired.kt

```diff
@@ -0,0 +1,120 @@
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.saveable.rememberSaveable
+import androidx.compose.runtime.setValue
+
+/**
+ * Composable that exercises the permissions flows as described in the
+ * [documentation](https://developer.android.com/training/permissions/requesting#workflow_for_requesting_permissions).
+ *
+ * @param permissionState required permission to be granted.
+ * @param permissionsGrantedContent content to show when the permission is granted.
+ * @param permissionsRationaleContent content to show when the user needs to be presented with
+ * a rationale for the required permission.
+ * @param permissionsDeniedContent content to show when the user denied the permission.
+ * @param permissionsRequestedContent content to show while the permission is being requested.
+ */
+@ExperimentalPermissionsApi
+@Composable
+fun PermissionRequired(
+    permissionState: PermissionState,
+    permissionsGrantedContent: @Composable (() -> Unit),
+    permissionsRationaleContent: @Composable (() -> Unit),
+    permissionsDeniedContent: @Composable (() -> Unit),
+    permissionsRequestedContent: @Composable (() -> Unit)? = null
+) {
+    var launchPermissionRequest by rememberSaveable { mutableStateOf(false) }
+
+    when {
+        permissionState.hasPermission -> {
+            permissionsGrantedContent()
+        }
+        permissionState.shouldShowRationale -> {
+            permissionsRationaleContent()
+        }
+        !permissionState.permissionRequested -> {
+            if (permissionsRequestedContent != null) {
+                permissionsRequestedContent()
+            }
+            launchPermissionRequest = true
+        }
+        else -> {
+            permissionsDeniedContent()
+        }
+    }
+
+    if (launchPermissionRequest) {
+        LaunchedEffect(permissionState) {
+            permissionState.launchPermissionRequest()
+            launchPermissionRequest = false
+        }
+    }
+}
+
+/**
+ * Composable that exercises the permissions flows as described in the
+ * [documentation](https://developer.android.com/training/permissions/requesting#workflow_for_requesting_permissions).
+ *
+ * @param multiplePermissionsState required permissions to be granted.
+ * @param permissionsGrantedContent content to show when the permissions are granted.
+ * @param permissionsRationaleContent content to show when the user needs to be presented with
+ * a rationale for the required permissions.
+ * @param permissionsDeniedContent content to show when the user denied the permissions.
+ * @param permissionsRequestedContent content to show while the permissions are being requested.
+ */
+@ExperimentalPermissionsApi
+@Composable
+fun PermissionsRequired(
+    multiplePermissionsState: MultiplePermissionsState,
+    permissionsGrantedContent: @Composable (() -> Unit),
```

Comment:
Do you think it's worth to provide the specific permission that was granted/revoked/require_rational or since the caller should have access to the state can check manually?

Reply:
Yeah, that's the idea. As the caller must hold the state instance, they can be as granular/generic as they want. The `MultiplePermissionsState` has a list of `revokedPermissions` that can be used for this